### PR TITLE
feat: add constructors for `VectorObject2D` and `MomentumObject2D`

### LINKED ISSUE #88: feat: new constructor
Placeholder issue for the constructor work, main feature of 0.9.

### PR DESCRIPTION
**Pull request by henryiii**
Closes #88.

WIP, currently just 2D Object implemented.

### CODE REVIEW

File: src/vector/backends/object.py [hunk 1]
```diff
@@ -799,17 +830,47 @@ class MomentumObject2D(PlanarMomentum, VectorObject2D):
...
+        elif kwargs and azimuthal is None:
+            if set(kwargs) == {"px", "py"}:
+                self.azimuthal = AzimuthalObjectXY(kwargs["px"], kwargs["py"])
+            elif set(kwargs) == {"pt", "phi"}:
+                self.azimuthal = AzimuthalObjectRhoPhi(kwargs["pt"], kwargs["phi"])
+            else:
+                raise TypeError("invalid arguments, must be px=, py= or pt=, phi=")
+        else:
+            raise TypeError("must give Azimuthal if not giving keyword arguments")
+
```
Comment (Saransh-cpp) [inline src/vector/backends/object.py:866]: I went through the overloads of `vector.obj` and a user can indeed create a `MomentumObject` by specifying a mixture of momentum and generic coordinates. I think this should be removed and the momentum coordinate names must be converted to generic in the `__init__` method of `VectorObject` classes, which will then be inherited by `MomentumObject` classes.

File: src/vector/backends/object.py [hunk 2]
```diff
@@ -662,17 +681,29 @@ def from_rhophi(cls, rho: float, phi: float) -> "VectorObject2D":
...
+            self.azimuthal = azimuthal
+        elif kwargs and azimuthal is None:
+            if set(kwargs) == {"x", "y"}:
+                self.azimuthal = AzimuthalObjectXY(kwargs["x"], kwargs["y"])
+            elif set(kwargs) == {"rho", "phi"}:
+                self.azimuthal = AzimuthalObjectRhoPhi(kwargs["rho"], kwargs["phi"])
+            else:
+                raise TypeError("invalid arguments, must be x=, y= or rho=, phi=")
+        else:
+            raise TypeError("must give Azimuthal if not giving keyword arguments")
```
Comment (Saransh-cpp) [inline src/vector/backends/object.py:701]: Removing the `MomentumObject` constructor and changing the `momentum` coordinates to `generic` ones works!
```suggestion
    def __init__(
        self, azimuthal: typing.Optional[AzimuthalObject] = None, **kwargs: float
    ) -> None:
        for k, v in kwargs.copy().items():
            kwargs.pop(k)
            kwargs[_repr_momentum_to_generic.get(k, k)] = v

        if not kwargs and azimuthal is not None:
            self.azimuthal = azimuthal
        elif kwargs and azimuthal is None:
            if set(kwargs) == {"x", "y"}:
                self.azimuthal = AzimuthalObjectXY(kwargs["x"], kwargs["y"])
            elif set(kwargs) == {"rho", "phi"}:
                self.azimuthal = AzimuthalObjectRhoPhi(kwargs["rho"], kwargs["phi"])
            else:
                raise TypeError("invalid arguments, must be x=, y= or rho=, phi=")
        else:
            raise TypeError("must give Azimuthal if not giving keyword arguments")
```

**Comment by Saransh-cpp** [comment]
I think this PR should work now, given that we are fine with passing generic coordinates into `MomentumObject` classes. The same is under discussion for the shortcut constructors (`obj` here), which can be handled in another PR.

**Comment by henryiii** [comment]
Feel free to turn of the unreachable check for mypy if it triggers too much, and 1-2 type ignores need `[misc]` added.

**Comment by Saransh-cpp** [comment]
Done! I've also refactored out the type safety code from the `obj` constructor.

**Comment by henryiii** [comment]
I can't approve "my own" (according to GitHub) PR, so verbally approving here!

**Comment by Saransh-cpp** [comment]
Merged! Thanks for the review! 🎉

**Pull request merged**